### PR TITLE
DCOS-10981: Network section multi containers

### PR DESCRIPTION
**Pull request by Poltergeist**
based on ~~#1506  #1519~~. These have merged :)

### CODE REVIEW

**Comment by nLight** [comment]
Placeholder goes missing when "Assign automatically" with a previous input. JSON is ok though.

dec-08-2016 18-41-04

**Comment by nLight** [comment]
Something is fishy with hostPort. It keeps my input in JSON, but after several clicks on the checkbox it resets it to 0

dec-08-2016 18-45-32

**Comment by nLight** [comment]
`containerPort` should be integer according to the raml, otherwise we can have "0123" as port

**Comment by jfurrow** [comment]
I am seeing the same bugs that @nLight pointed out.

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 1]
```diff
@@ -154,12 +265,70 @@ module.exports = {
...
+          break;
+        case REMOVE_ITEM:
+          this.endpoints[path[1]].endpoints = this.endpoints[path[1]].endpoints.filter((item, index) => {
+            return index !== value;
+          });
+          break;
+      }
+
+      if (type === SET) {
+        if (path[4] === 'name') {
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: dude!
```js
const [base, index, field, secondIndex, name] = path;
const names = ['name', 'hostPort', 'containerPort', 'protocol', 'automaticPort', 'portMapping', 'loadBalanced', 'vip'];
if (type === SET && names.includes(name)) {
  this.endpoints[index].endpoints[secondIndex][name] = value;
}

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 2]
```diff
@@ -261,6 +430,53 @@ module.exports = {
...
+        case REMOVE_ITEM:
+          newState[path[1]].endpoints = newState[path[1]].endpoints.filter((item, index) => {
+            return index !== value;
+          });
+          break;
+      }
+
+      if (type === SET) {
+        if (path[4] === 'name') {
+          newState[path[1]].endpoints[path[3]].name = value;
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: Same deal here

File: plugins/services/src/js/reducers/serviceForm/MultiContainerNetwork.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      return new Transaction(['network'], `${mode}.${name}`);
+    }
+
+    return new Transaction(['network'], state.network.mode.toUpperCase());
+  },
+  FormReducer(state = {mode: Networking.type.HOST}, {type, path, value}) {
+    const joinedPath = path.join('.');
+
+    if (joinedPath === 'network' && type === SET) {
+      const valueSplit = value.split('.');
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerNetwork.js]: `const [mode, name] = value.split('.');`?

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 4]
```diff
@@ -101,6 +148,47 @@ function containersParser(state) {
       });
     }
 
+    if (item.endpoints != null && item.endpoints.length !== 0) {
+      item.endpoints.forEach((endpoint, endpointIndex) => {
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: Could we break this out into separate functions?
The nesting is hard to follow

File: plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+import HostUtil from '../../utils/HostUtil';
+import Icon from '../../../../../../src/js/components/Icon';
+import Networking from '../../../../../../src/js/constants/Networking';
+import VirtualNetworksStore from '../../../../../../src/js/stores/VirtualNetworksStore';
+
+const METHODS_TO_BIND = [
+  'onVirtualNetworksStoreSuccess'
+];
+
+class MultiContainerHealthChecksFormSection extends mixin(StoreMixin) {
```
Comment (mlunoe) [inline plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js]: Let's try to create section parts we can re-use and then have one section instead of copying what looks to be almost the same as `NetworkingFormSection`

File: plugins/services/src/js/reducers/serviceForm/Containers.js (hunk 1 above)
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: done!

File: plugins/services/src/js/reducers/serviceForm/Containers.js (hunk 2 above)
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: done!

File: plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js (hunk 5 above)
Comment (nLight) [inline plugins/services/src/js/components/forms/MultiContainerNetworkingFormSection.js]: Fixed copy-paste `HealthChecks`, didn't create reusable section parts.

File: plugins/services/src/js/reducers/serviceForm/Containers.js (hunk 4 above)
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: Yes and no, sorry, not now

**Comment by nLight** [comment]
I addressed all the comments except 1 because there's a tiny time pressure, let's get this in

**Comment by nLight** [comment]
@Poltergeist would you please have a quick look and merge at will?

**Comment by orlandohohmeier** [comment]
🚀 ✨ 🚢
